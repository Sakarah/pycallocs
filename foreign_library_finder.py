import allocs
import importlib
import importlib.abc
import importlib.machinery
import sys

# These can be modified by the user to change the finder search paths
lib_extension = ".so"

# The blank search path has a special meaning as dlopen is using its own search 
# paths when the file name does not contain any /
search_paths = [""]

class ForeignLibraryFinder(importlib.abc.MetaPathFinder):
    """
        Meta path finder searching foreign libraries usable
        with allocs.ForeignLibraryLoader.
    """

    def find_spec(fullname, path, target):
        if path is not None:
            # Should we support submodules ?
            return None
        
        for base_path in search_paths:
            try:
                filename = base_path + fullname + lib_extension
                loader = allocs.ForeignLibraryLoader(filename)
                return importlib.machinery.ModuleSpec(fullname, loader, origin=filename)
            except ImportError:
                continue

        return None

sys.meta_path.append(ForeignLibraryFinder)
