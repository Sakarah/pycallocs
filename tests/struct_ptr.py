import foreign_library_finder
foreign_library_finder.search_paths.append("libs/")
import composite
import bintree

hw = composite.hello_world(666)
composite.compl_hw(hw)
print(hw)

bt = bintree.bintree(5)
bintree.bst_insert(bt, 1)
bintree.bst_insert(bt, 44)
bintree.bst_insert(bt, -8)
bintree.bst_insert(bt, 3)
bintree.bst_insert(bt, 98)
bintree.bst_insert(bt, 8)
print(bt)
bintree.print_bintree(bt)
