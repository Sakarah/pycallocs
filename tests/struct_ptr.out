(hello_world){hello: 666, world: 666.0}
(bintree){val: 5, left: (bintree){val: 1, left: (bintree){val: -8, left: None, right: None}, right: (bintree){val: 3, left: None, right: None}}, right: (bintree){val: 44, left: (bintree){val: 8, left: None, right: None}, right: (bintree){val: 98, left: None, right: None}}}
{{{-8}1{3}}5{{8}44{98}}}